File: src/udemy_tutorials_cpp/src/smartphone.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"


class SmartphoneNode: public rclcpp::Node
{
    public:
        SmartphoneNode(): Node("smartphone")
        {
            // Create a subscriber on the topic "robot_news"
            subscriber_ = this->create_subscription<example_interfaces::msg::String>(
                "robot_news",
                10,  // Queue size
                std::bind(&SmartphoneNode::callback_robot_news,
                          this,
                          std::placeholders::_1)
            );
            RCLCPP_INFO(this->get_logger(), "Smartophone has been started.");
        }
    private:
        void callback_robot_news(const example_interfaces::msg::String::SharedPtr msg)
        {
            RCLCPP_INFO(
                this->get_logger(),
                "Message received: '%s'",
                msg->data.c_str()
            );
        }
        rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;
};


int main(int argc, char **argv)
{
    // Initialize ROS2
    rclcpp::init(argc, argv);
    // Create a SmartphoneNode
    auto node = std::make_shared<SmartphoneNode>();
    // Spin the node
    rclcpp::spin(node);
    // Shut down ROS2
    rclcpp::shutdown();
    return 0;
}

File: src/udemy_tutorials_py/udemy_tutorials_py/smartphone.py
#!/usr/bin/env python3
"""Example of ROS2 subscriber.

This scripts creates an 'smartphone' node which subscribes to the
'robot_news_station' topic and prints the messages to the console.

"""
import sys
import rclpy
from rclpy.node import Node
from example_interfaces.msg import String


class SmartphoneNode(Node):
    """Smartphone node class."""
    def __init__(self):
        super().__init__('smartphone')
        # Create subscriber
        self.subscriber = self.create_subscription(
            msg_type=String,
            topic='robot_news',
            callback=self.callback_robot_news,
            qos_profile=10,  # Queue size (Same as robot_news_station)
        )
        # Show message that node is running
        self.logger.info('Smartphone node been started.')

    @property
    def logger(self):
        """Return logger."""
        return self.get_logger()

    def callback_robot_news(self, msg):
        """Handle received messages."""
        self.logger.info(f"Message received: {msg.data}")


def main(args=None):
    """Main function."""
    rclpy.init(args=args)
    smartphone_node = SmartphoneNode()
    rclpy.spin(smartphone_node)
    rclpy.shutdown()
    return 0


if __name__ == '__main__':
    sys.exit(main())
